File: cli/core/vectors_test.go
// Harness dos vetores de conformidade (spec/SPEC.md §6) — mesma bateria que
// roda no cliente JS (extension/test/vectors.test.js): asserção por inclusão
// parcial e reexecução com a entrada embaralhada quando "permutations": true.
package core_test

import (
	"encoding/json"
	"fmt"
	"math/rand/v2"
	"os"
	"path/filepath"
	"reflect"
	"testing"

	"github.com/onofre-matheus/tcc/cli/core"
)

const permutationRuns = 5

type vector struct {
	Name         string          `json:"name"`
	Projection   string          `json:"projection"`
	Now          string          `json:"now"`
	TZ           string          `json:"tz"`
	Permutations bool            `json:"permutations"`
	Events       []core.Event    `json:"events"`
	Expected     json.RawMessage `json:"expected"`
}

func TestVectors(t *testing.T) {
	files, err := filepath.Glob(filepath.Join("..", "..", "spec", "vectors", "*.json"))
	if err != nil {
		t.Fatal(err)
	}
	if len(files) == 0 {
		t.Fatal("nenhum vetor encontrado em spec/vectors")
	}

	for _, file := range files {
		raw, err := os.ReadFile(file)
		if err != nil {
			t.Fatal(err)
		}
		var v vector
		if err := json.Unmarshal(raw, &v); err != nil {
			t.Fatalf("%s: %v", file, err)
		}

		t.Run(filepath.Base(file), func(t *testing.T) {
			fn, ok := core.Lookup(v.Projection)
			if !ok {
				t.Fatalf("projeção não registrada: %q", v.Projection)
			}

			run := func(t *testing.T, events []core.Event) {
				got, err := fn(events, core.Params{Now: v.Now, TZ: v.TZ})
				if err != nil {
					t.Fatal(err)
				}
				assertMatch(t, asJSON(t, got), asJSON(t, v.Expected), "$")
			}

			t.Run("produz o estado esperado", func(t *testing.T) {
				run(t, v.Events)
			})

			if v.Permutations {
				for seed := uint64(1); seed <= permutationRuns; seed++ {
					t.Run(fmt.Sprintf("converge com a entrada embaralhada (permutação %d)", seed), func(t *testing.T) {
						events := append([]core.Event(nil), v.Events...)
						rng := rand.New(rand.NewPCG(seed, 0))
						rng.Shuffle(len(events), func(i, j int) {
							events[i], events[j] = events[j], events[i]
						})
						run(t, events)
					})
				}
			}
		})
	}
}

// asJSON normaliza qualquer valor para a forma genérica de JSON decodificado
// (map[string]any, []any, float64, string, bool, nil), para que a comparação
// não dependa dos tipos concretos das projeções.
func asJSON(t *testing.T, v any) any {
	t.Helper()
	raw, err := json.Marshal(v)
	if err != nil {
		t.Fatal(err)
	}
	var out any
	if err := json.Unmarshal(raw, &out); err != nil {
		t.Fatal(err)
	}
	return out
}

// assertMatch aplica a inclusão parcial da spec: todo campo presente em
// expected deve ter valor idêntico; campos ausentes são livres; arrays são
// comparados por igualdade exata.
func assertMatch(t *testing.T, actual, expected any, path string) {
	t.Helper()
	switch exp := expected.(type) {
	case []any:
		if !reflect.DeepEqual(actual, exp) {
			t.Fatalf("%s: esperado %v, veio %v", path, exp, actual)
		}
	case map[string]any:
		act, ok := actual.(map[string]any)
		if !ok {
			t.Fatalf("%s: esperado objeto, veio %T (%v)", path, actual, actual)
		}
		for key, val := range exp {
			assertMatch(t, act[key], val, path+"."+key)
		}
	default:
		if !reflect.DeepEqual(actual, expected) {
			t.Fatalf("%s: esperado %v (%T), veio %v (%T)", path, expected, expected, actual, actual)
		}
	}
}

File: extension/test/vectors.test.js
import { describe, it, expect } from "vitest";
import { readdirSync, readFileSync } from "node:fs";
import { join, dirname } from "node:path";
import { fileURLToPath } from "node:url";
import { project as leitner } from "../core/leitner.js";
import { attentionWindowSeconds } from "../core/attention.js";
import { project as tasks } from "../core/tasks.js";
import { project as decks } from "../core/decks.js";
import { project as stats } from "../core/stats.js";
import { project as inbox } from "../core/inbox.js";
import { project as calendar, reminders } from "../core/calendar.js";
import { project as review } from "../core/review.js";

const vectorsDir = join(
  dirname(fileURLToPath(import.meta.url)),
  "..", "..", "spec", "vectors"
);

const PERMUTATION_RUNS = 5;

function runProjection(vector, events) {
  switch (vector.projection) {
    case "leitner":
      return leitner(events, { now: vector.now, tz: vector.tz });
    case "attention":
      return { attention_seconds: attentionWindowSeconds(events) };
    case "tasks":
      return tasks(events);
    case "decks":
      return decks(events);
    case "stats":
      return stats(events, { now: vector.now, tz: vector.tz });
    case "inbox":
      return inbox(events);
    case "calendar":
      return calendar(events, { now: vector.now, tz: vector.tz });
    case "reminders":
      return reminders(events, { now: vector.now, tz: vector.tz });
    case "review":
      return review(events, { now: vector.now, tz: vector.tz });
    default:
      throw new Error(`projeção desconhecida: ${vector.projection}`);
  }
}

// Asserção por inclusão parcial (ver spec/SPEC.md §6): todo campo presente em
// expected deve bater; arrays são comparados por igualdade exata.
function expectMatch(actual, expected, path = "$") {
  if (Array.isArray(expected)) {
    expect(actual, path).toEqual(expected);
    return;
  }
  if (expected !== null && typeof expected === "object") {
    expect(actual, path).toBeTypeOf("object");
    for (const key of Object.keys(expected)) {
      expectMatch(actual?.[key], expected[key], `${path}.${key}`);
    }
    return;
  }
  expect(actual, path).toBe(expected);
}

// Embaralhamento determinístico (mulberry32) para os vetores de convergência.
function shuffled(array, seed) {
  let s = seed >>> 0;
  const rand = () => {
    s = (s + 0x6d2b79f5) >>> 0;
    let t = Math.imul(s ^ (s >>> 15), 1 | s);
    t = (t + Math.imul(t ^ (t >>> 7), 61 | t)) ^ t;
    return ((t ^ (t >>> 14)) >>> 0) / 4294967296;
  };
  const copy = [...array];
  for (let i = copy.length - 1; i > 0; i--) {
    const j = Math.floor(rand() * (i + 1));
    [copy[i], copy[j]] = [copy[j], copy[i]];
  }
  return copy;
}

const files = readdirSync(vectorsDir).filter((f) => f.endsWith(".json")).sort();

describe("vetores de conformidade (spec/vectors)", () => {
  for (const file of files) {
    const vector = JSON.parse(readFileSync(join(vectorsDir, file), "utf8"));

    describe(`${file}: ${vector.name}`, () => {
      it("produz o estado esperado", () => {
        expectMatch(runProjection(vector, vector.events), vector.expected);
      });

      if (vector.permutations) {
        for (let seed = 1; seed <= PERMUTATION_RUNS; seed++) {
          it(`converge com a entrada embaralhada (permutação ${seed})`, () => {
            const events = shuffled(vector.events, seed);
            expectMatch(runProjection(vector, events), vector.expected);
          });
        }
      }
    });
  }
});
